assembly/src/parsers/u32_ops.rs

```diff
@@ -186,19 +189,74 @@ pub fn parse_u32xor(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
     Ok(())
 }
 
-/// TODO: implement
-pub fn parse_u32not(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Translates u32not assembly instruction to VM operation PAD U32OR SWAP U32XOR.
+pub fn parse_u32not(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => {
+            span_ops.push(Operation::Push(BaseElement::new(2 ^ 32)));
+            span_ops.push(Operation::Swap);
+            span_ops.push(Operation::Incr);
+            span_ops.push(Operation::U32sub);
+        }
+        _ => return Err(AssemblyError::extra_param(op)),
+    }
+
+    Ok(())
 }
 
-/// TODO: implement
-pub fn parse_u32shl(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Translates u32shl.x assembly instruction to VM operation PUSH(2^x) MUL U32DIV DROP.
+pub fn parse_u32shl(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0..=1 => return Err(AssemblyError::missing_param(op)),
+        2 => {
+            // Assert the value is u32
+            span_ops.push(Operation::U32split);
+            span_ops.push(Operation::Not);
+            span_ops.push(Operation::Assert);
+
+            match parse_int_param(op, 1, 1, 31) {
+                Ok(x) => {
+                    let base: u64 = 2;
+                    span_ops.push(Operation::Push(BaseElement::new(base.pow(x))));
+                    span_ops.push(Operation::Mul);
+                    span_ops.push(Operation::U32div);
+                    span_ops.push(Operation::Drop);
+                }
+                Err(e) => return Err(e),
```

Comment:
We can get rid of `match` in this (and other similar blocks) as:

```Rust
let x = parse_int_param(op, 1, 1, 31)?;
span_ops.push(Operation::Push(BaseElement::new(2u64.pow(x))));
span_ops.push(Operation::Mul);
span_ops.push(Operation::U32div);
span_ops.push(Operation::Drop);
```

Reply:
Ah yes! Still getting familiar with all the rust goodness :)